### PR TITLE
Life of dist txn

### PR DESCRIPTION
**Pull request by sploiselle**
@nvanbenschoten Could I get you to take a look at this for technical accuracy? Still need to polish some of the writing, but any other copy edits are certainly welcome.

### CODE REVIEW

File: v2.2/architecture/life-of-a-distributed-transaction.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+The gateway node handles the connection with the client, both receiving and responding to the request.
+
+### Parsing
+
+The gateway node first [parses](sql-layer.html#sql-parser-planner-executor) the client's SQL statement to ensure it's valid according to the CockroachDB dialect of PostgreSQL.
+
+While CockroachDB presents a SQL interface to clients, the actual database is built on top of a key-value store. Parsed SQL statements are converted into key-value operations, and a the node [generates a plan](sql-layer.html#logical-planning) to execute the query.
+
+### TxnCoordSender
```
Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: I'd make it clear that this lives on the gateway node.

File: v2.2/architecture/life-of-a-distributed-transaction.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+The gateway node's `DistSender` receives `BatchRequests` from the `TxnCoordSender`. It dismantles the initial `BatchRequest` by taking each operation and finding which physical machine should receive the request for the range––known as the range's leaseholder. The address of the range's current leaseholder is readily available in both local caches, as well as in the [cluster's `meta` ranges](distribution-layer.html#meta-range-kv-structure).
+
+These dismantled `BatchRequests` are reassembled into new `BatchRequests` containing the address of the range's leaseholder.
+
+All writes operations also propagate the leaseholder's address back to the `TxnCoordSender`, so it can track and clean up write operations as necessary.
+
+The `DistSender` sends out the first `BatchRequest` for each range in parallel. As soon as it receives a provisional acknowledgment from the leasholder node’s evaluator (details below), it sends out the next `BatchRequest` for that range.
+
+The DistSender then waits to receive commit acknowledgments for all of its write operations, as well as values for all of its read operations.
```
Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: Consider removing "commit" from the term "commit acknowledgments". The reason for this is that commit means different things in different layers of CockroachDB. The DistSender might be performing an operation within an ongoing SQL transaction and will still wait for an acknowledgment.

File: v2.2/architecture/life-of-a-distributed-transaction.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+### Request response
+
+Because the leaseholder replica can shift between nodes, all nodes must be able to return a request for any key.
+
+##### No Longer Leaseholder
+
+If a node:
+
+- *Was* the leaseholder, but no longer holds the lease
+- Holds a replica for the range
```
Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: Does not hold a replica for the range? I'm not sure what you're trying to enumerate here.

File: v2.2/architecture/life-of-a-distributed-transaction.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+If a node doesn't have a replica for the requested range, it denies the request without providing any further information.
+
+In this case, the `DistSender` must look up the current leaseholder using the [cluster's `meta` ranges](distribution-layer.html#meta-range-kv-structure) range.
+
+##### Success
+
+Once the node that contains the leaseholder of the range receives the `BatchRequest`, it begins processing it, and progresses onto checking the timestamp cache.
+
+### Timestamp cache
```
Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: Consider making a note that this is how Cockroach ensures serializability.

File: v2.2/architecture/life-of-a-distributed-transaction.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+Each write operation in a `BatchRequest` checks its own timestamp versus the timestamp cache to ensure that the write operation has a higher timestamp; this guarantees that history is never rewritten and you can trust that reads always served the most recent data. If a write operation fails this check, it must be restarted at a timestamp higher than the timestamp cache's value.
+
+### Latch manager
+
+Operations in the `BatchRequest` are serialized through the leaseholder's latch manager.
+
+This works by giving each write operation a latch on a row. Any reads or writes that come in after the latch has been granted on the row must wait for the write to complete, at which point the latch is released and the subsequent operations can continue.
+
+### Evaluator
```
Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: I'm not familiar with this term in the storage layer, although there is a SQL evaluator, which I think you're referring to. `CHECK` constraints are validated at that level, which is up above the `TxnCoordSender`.

Comment (nvanbenschoten) [reply v2.2/architecture/life-of-a-distributed-transaction.md]: You might also be referring to "batch evaluation", which is where `BatchRequests` determine their result under the "protection" of the latch manager. I'd suggest renaming this to `Batch Evaluation` and removing any reference to `CHECK` cosntraints.

File: v2.2/architecture/life-of-a-distributed-transaction.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+What we detail below is a simplified version of the CockroachDB transaction model. For more detail, check out [the transaction architecture documentation](transaction-layer.html).
+
+#### Resolving Write Intents
+
+If an operation encounters a write intent for a key, it attempts to "resolve" the write intent by checking the write intent's transaction's state. If the write's intent's transaction is...
+
+- `COMMITTED`, this operation converts the write intent to a regular key-value pair, and then proceeds as if it had read that value instead of a write intent.
+- `ABORTED`, this operation discards the write intent and reads the next-most-recent value from RocksDB.
+- `PENDING`, the new transaction attempts to "push" the write intent's transaction by moving that transaction's timestamp forward (i.e. ahead of this transaction's timestmap). The push succeeds only if this transaction's priority is explicitly higher; i.e. it has a `HIGH` priority or the transaction whose write intent it encountered has a `LOW` priority.
```
Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: It's no longer sufficient for the transaction's priority to be explicitly higher. Instead, it must either have the maximum priority or the other txn must have the minimum priority for the push to succeed. SQL-related txns will never have these extreme priorities, so they'll never result in pushes succeeding. Instead, transactions will queue up behind each other.

This is part of a shift we made back in 1.1 to move from an "optimistic" concurrency control to a "pessimistic" concurrency control. We observed that in a distributed environment where txn aborts are very expensive, OCC was too costly.

File: v2.2/architecture/life-of-a-distributed-transaction.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+As we mentioned before, each read operation also updates the timestamp cache.
+
+### Write Operations
+
+After guaranteeing that there are no existing write intents for the keys, `BatchRequest`'s key-value operations are converted to [Raft operations](replication-layer.html#raft) and have their values converted into write intents.
+
+The leaseholder then proposes these Raft operations to the Raft group leader. Most often, the leaseholder and the Raft leader are the same node, but there are situations where the roles might drift to different nodes. However, when the two roles are not collocated on the same physical machine, CockroachDB will attempt to relocate them on the same node at the next opportunity.
+
+## Raft Leader
```
Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: I'd make a note that the Range leaseholder and that Range's Raft leader are almost always the same node.

File: v2.2/architecture/life-of-a-distributed-transaction.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+Note that this write operation creates a write intent; these writes will not be fully committed until the gateway node's `TxnCoordSender` sets the transaction record's status to `COMMITTED`.
+
+## On the way back up
+
+Now that we have followed an operation all the way down from the SQL client to RocksDB, we can pretty quickly cover what happens on the way back up (i.e., generating a response to the client).
+
+1. Once the leaseholder applies a write to its Raft log, it sends an commit acknowledgment to the gateway node's `DistSender`, which was waiting for this signal (having already received the provisional acknowledgment from the leaseholder's evaluator).
+1. The gateway node's `DistSender` aggregates commit acknowledgments from all of the write operations in the `BatchRequest`, as well as any values from read operations that should be returned to the client.
+1. Once all operations have successfully completed, the `DistSender` sends a response to the `TxnCoordSender`, indicating that the transaction completed.
+1. The `TxnCoordSender` checks the transaction record. There are a few situations that might arise here:
```
Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: This is confusing to me. I think the reason why is that it's not clear throughout this piece whether we're referring to the lifetime of a single SQL statement outside of a transaction, a single SQL statement inside a transaction, or multiple SQL statements inside a transaction. Are we trying to commit the transaction at all?

Comment (nvanbenschoten) [inline v2.2/architecture/life-of-a-distributed-transaction.md]: Also, note that the `TxnCoordSender` never does anything itself. It always operates by sending `BatchRequests` down through a `DistSender`.

**Comment by sploiselle** [comment]
@nvanbenschoten Could I get a quick look over on a few final things?

- [SQL planning & parsing; executor sections](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f2fb52f7dc716a3468bc4e8d277dc73795bdb355/dev/architecture/life-of-a-distributed-transaction.html#sql-parsing-planning)
- The following line from [On the way back up](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f2fb52f7dc716a3468bc4e8d277dc73795bdb355/dev/architecture/life-of-a-distributed-transaction.html#sql-parsing-planning):
    > Once all operations have successfully completed, the `DistSender` checks the transaction record. There are a few situations that might arise here:

Thanks so much for your help with this.

**Comment by sploiselle** [comment]
@jseldess @rmloveland Could I get a copy edit on this from either of you?

**Comment by rmloveland** [comment]
> @rmloveland Sorry you had to clean up so much mess in here

@sploiselle No worries mate, they're just typos.  We all make them here in cornputer world :-}

PS Thanks for answering my Q about which errors bubble up to the SQL interface.  Sounds like an area for future docs work, probably.

**Pull request merged**